worker/src/main/java/gov/cms/ab2d/worker/config/RoundRobinBlockingQueue.java

```diff
@@ -0,0 +1,357 @@
+        }
+    }
+
+    @Override
+    public boolean retainAll(@NotNull Collection<?> c) {
+        throw new UnsupportedOperationException("Complicated and probably not needed so unless it is, not implementing");
+    }
+
+    public int size() {
+        return (int) count;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return count == 0;
+    }
+
+    public E take() throws InterruptedException {
+        final ReentrantLock lock = this.lock;
+        lock.lockInterruptibly();
+        try {
+            while (count == 0) {
+                notEmpty.await();
+            }
+            return getNext();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object[] toArray() {
+        throw new UnsupportedOperationException("Complicated and probably not needed so unless it is, not implementing");
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        throw new UnsupportedOperationException("Complicated and probably not needed so unless it is, not implementing");
+    }
+
+    public String toString() {
+        if (categories.isEmpty()) {
+            return "Empty";
+        }
+        StringBuilder builder = new StringBuilder();
+        for (String category : categories) {
+            builder.append(category).append(" with ").append(categoryQueues.get(category).size()).append(" Futures; ");
+        }
+        return builder.toString();
+    }
+
+    /**
+     * Get the next round robin Object
+     *
+     * @return
+     */
+    private E getNext() {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            if (categoryQueues == null || categories.isEmpty()) {
+                return null;
+            }
+            if (currentIndex >= categories.size()) {
+                currentIndex = 0;
+            }
+            String currentContract = categories.get(currentIndex);
+            Queue<E> currentFuture = categoryQueues.get(currentContract);
+            E val = currentFuture.poll();
+            count--;
+            if (currentFuture.size() == 0) {
+                // No more requests so remove category. We don't have to increment index since we removed the item at the
+                // current index
+                categoryQueues.remove(currentFuture);
+                categories.remove(currentContract);
+            } else {
+                currentIndex++;
+            }
+            return val;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private E peekNext() {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            if (categoryQueues == null || categories.isEmpty()) {
+                return null;
+            }
+            if (currentIndex >= categories.size()) {
+                currentIndex = 0;
+            }
+            String currentContract = categories.get(currentIndex);
+            return categoryQueues.get(currentContract).peek();
```

Comment:
We should only do `peek` on the current queue is it is not empty. Otherwise, `peek` returns null, and `peekNext` returns null, too, while there may be other queues in rotation that are not empty.

Reply:
...unless of course I have missed a spot that ensures we never have empty queues in rotation.

Reply:
Yes, I test to see if the queue is empty when I remove an item. If it is, I remove that list and the "category", aka contract from the contact list.

            if (linkedList.size() == 0) {
                categoryQueues.remove(entry.getKey());
                categories.remove(entry.getKey());
            }